File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/serialization/NetworkSerialization.java
package net.namekdev.entity_tracker.utils.serialization;

import com.artemis.utils.BitVector;

public abstract class NetworkSerialization {
	public static enum Type {
		Unknown,
		Null,
		Array,
		Description,
		MultipleDescriptions,
		DescriptionRef,
		Object,
		Enum,
		EnumDescription,
		EnumValue,
		
		Byte,
		Short,
		Int,
		Long,
		String,
		Boolean, //takes 1 byte
		Float,
		Double,
		BitVector //takes minimum 4 bytes
	}

	public static NetworkSerializer createSerializer() {
		return new NetworkSerializer();
	}

	public static NetworkDeserializer createDeserializer() {
		return new NetworkDeserializer();
	}

	public static Type determineType(Class<?> type) {
		Type netType = Type.Unknown;

		if (type.equals(byte.class) || type.equals(Byte.class))
			netType = Type.Byte;
		else if (type.equals(short.class) || type.equals(Short.class))
			netType = Type.Short;
		else if (type.equals(int.class) || type.equals(Integer.class))
			netType = Type.Int;
		else if (type.equals(long.class) || type.equals(Long.class))
			netType = Type.Long;
		else if (type.equals(String.class))
			netType = Type.String;
		else if (type.equals(boolean.class) || type.equals(Boolean.class))
			netType = Type.Boolean;
		else if (type.equals(float.class) || type.equals(Float.class))
			netType = Type.Float;
		else if (type.equals(double.class) || type.equals(Double.class))
			netType = Type.Double;
		else if (type.equals(BitVector.class))
			netType = Type.BitVector;
		else if (type.isEnum())
			netType = Type.Enum;

		return netType;
	}

	public static Object convertStringToTypedValue(String value, Type valueType) {
		switch (valueType) {
			case Byte: return Byte.valueOf(value);
			case Short: return Short.valueOf(value);
			case Int: return Integer.valueOf(value);
			case Long: return Long.valueOf(value);
			case String: return value;
			case Boolean: return Boolean.valueOf(value);
			case Float: return Float.valueOf(value);
			case Double: return Double.valueOf(value);
			case Enum: throw new UnsupportedOperationException("probably unsupported, not sure");
			case BitVector: return new BitVector(Integer.valueOf(value));
			case Array: throw new UnsupportedOperationException("arrays are not supported (yet?)");
			default: return null;
		}
	}

	public static boolean isSimpleType(Type valueType) {
		if (valueType == null)
			return false;

		switch (valueType) {
			case Byte: return true;
			case Short: return true;
			case Int: return true;
			case Long: return true;
			case String: return true;
			case Boolean: return true;
			case Float: return true;
			case Double: return true;
			case BitVector: return true;
			default: return false;
		}
	}
}

File: artemis-entity-tracker/src/main/java/net/namekdev/entity_tracker/utils/serialization/NetworkSerialization.kt
package net.namekdev.entity_tracker.utils.serialization

import com.artemis.utils.BitVector

abstract class NetworkSerialization {
    enum class DataType {
        Undefined,

        // meta data
        Description,
        MultipleDescriptions,
        EnumDescription,  //defines a list of possible values in enum
        EnumValue,

        // simple timpes
        Byte,
        Short,
        Int,
        Long,
        String,
        Boolean, //takes 1 byte
        Float,
        Double,
        BitVector, //takes minimum 4 bytes, BitVector is a type from artemis-odb

        // more complicated data types
        Unknown,
        Object,
        Array,
        Enum,

        // special values
        Null, //takes 1 byte
        DescriptionRef,
    }


    companion object {
        fun createSerializer(): NetworkSerializer {
            return NetworkSerializer()
        }

        fun createDeserializer(): NetworkDeserializer {
            return NetworkDeserializer()
        }

        /**
         * @return pair of dataType + isTypePrimitive
         */
        fun determineType(type: Class<*>): Pair<DataType, Boolean> {
            return when (type) {
                Byte::class.javaPrimitiveType -> Pair(DataType.Byte, true)
                Byte::class.javaObjectType -> Pair(DataType.Byte, false)
                Short::class.javaPrimitiveType -> Pair(DataType.Short, true)
                Short::class.javaObjectType -> Pair(DataType.Short, false)
                Int::class.javaPrimitiveType -> Pair(DataType.Int, true)
                Int::class.javaObjectType -> Pair(DataType.Int, false)
                Long::class.javaPrimitiveType -> Pair(DataType.Long, true)
                Long::class.javaObjectType -> Pair(DataType.Long, false)
                String::class.java -> Pair(DataType.String, false)
                Boolean::class.javaPrimitiveType -> Pair(DataType.Boolean, true)
                Boolean::class.javaObjectType -> Pair(DataType.Boolean, false)
                Float::class.javaPrimitiveType -> Pair(DataType.Float, true)
                Float::class.javaObjectType -> Pair(DataType.Float, false)
                Double::class.javaPrimitiveType -> Pair(DataType.Double, true)
                Double::class.javaObjectType -> Pair(DataType.Double, false)
                BitVector::class.java -> Pair(DataType.BitVector, false)
                else ->
                    if (type.isEnum)
                        Pair(DataType.Enum, false)
                    else
                        Pair(DataType.Unknown, false)
            }
        }

        fun convertStringToTypedValue(value: String, valueType: DataType): Any? {
            when (valueType) {
                DataType.Byte -> return java.lang.Byte.valueOf(value)
                DataType.Short -> return java.lang.Short.valueOf(value)
                DataType.Int -> return Integer.valueOf(value)
                DataType.Long -> return java.lang.Long.valueOf(value)
                DataType.String -> return value
                DataType.Boolean -> return java.lang.Boolean.valueOf(value)
                DataType.Float -> return java.lang.Float.valueOf(value)
                DataType.Double -> return java.lang.Double.valueOf(value)
                DataType.Enum -> throw UnsupportedOperationException("probably unsupported, not sure")
                DataType.BitVector -> return BitVector(Integer.valueOf(value)!!)
                else -> return null
            }
        }

        fun isSimpleType(valueType: DataType?): Boolean {
            if (valueType == null)
                return false

            when (valueType) {
                DataType.Byte -> return true
                DataType.Short -> return true
                DataType.Int -> return true
                DataType.Long -> return true
                DataType.String -> return true
                DataType.Boolean -> return true
                DataType.Float -> return true
                DataType.Double -> return true
                DataType.BitVector -> return true
                else -> return false
            }
        }
    }
}
